### ISSUE TITLE
how to save segmented image annotation in yolo format ? or segmentation support format

### ISSUE THREAD

**Issue by akashAD98** [facebookresearch/segment-anything#215]
i tested it on image & im getting the mask, now i want to save this mask/annotations in yolo or mask rcnn or any segmentation format.
what should i need to do.
thanks.

<img width="421" alt="image" src="https://user-images.githubusercontent.com/62583018/232119518-f1fbef4b-e3d4-4ef0-8a9f-6102e8d2b298.png">

**Comment by Jordan-Pierce** [comment]
If you're doing instance segmentation using COCO format, you'd just need to provide the bounding box output from SAM model for the given mask, and for the instance segmentation, you'd probably need to use something like OpenCv's find contour method to get a list of the vertices, and supply to the segmentation variable in the file. Alternatively, given the mask, just calculate the bounding box coordinates:

```python
# Find contours in the mask image
contours, _ = cv2.findContours(mask, cv2.RETR_EXTERNAL, cv2.CHAIN_APPROX_NONE)

# Convert the contour to the format required for segmentation in COCO format
segmentation = []
for contour in contours:
    # Flatten the contour coordinates into a 1D array
    contour = contour.flatten().tolist()
    # Convert the 1D array to a list of (x, y) pairs
    contour_pairs = [(contour[i], contour[i+1]) for i in range(0, len(contour), 2)]
    # Convert the (x, y) pairs to a list of integers
    segmentation.append([int(coord) for pair in contour_pairs for coord in pair])

# Define annotation in COCO format
annotation = {
    'id': 1,
    'image_id': 1,
    'category_id': 1,
    'segmentation': segmentation,
    'area': int(cv2.contourArea(contours[0])),
    'bbox': [int(x) for x in cv2.boundingRect(contours[0])],
    'iscrowd': 0
}
```

COCO format:
```
{
    "info": {},
    "licenses": [],
    "images": [
        {
            "id": 1,
            "file_name": "example.jpg",
            "height": 480,
            "width": 640
        }
    ],
    "annotations": [
        {
            "id": 1,
            "image_id": 1,
            "category_id": 1,
            "segmentation": [[10,10,100,10,100,100,10,100]],
            "area": 8100,
            "bbox": [10,10,90,90],
            "iscrowd": 0
        }
    ],
    "categories": [
        {
            "id": 1,
            "name": "example",
            "supercategory": "object"
        }
    ]
}
```

For YOLO format, you just modify the bbox coordinates and write a text file:
```python
# Write bounding boxes to file in YOLO format
with open('bounding_boxes.txt', 'w') as f:
    for contour in contours:
        # Get the bounding box coordinates of the contour
        x, y, w, h = cv2.boundingRect(contour)
        # Convert the coordinates to YOLO format and write to file
        f.write('0 {:.6f} {:.6f} {:.6f} {:.6f}\n'.format((x+w/2)/mask_image.shape[1], (y+h/2)/mask_image.shape[0], w/mask_image.shape[1], h/mask_image.shape[0]))
```
YOLO format:
```
0 0.25 0.25 0.5 0.5
```

**Comment by yeldarby** [comment]
We added support in Roboflow for [annotating using SAM](https://blog.roboflow.com/label-data-segment-anything-model-sam/). We convert the masks to polygons & you can export to YOLO format for training a model.

**Comment by akashAD98** [comment]
@Jordan-Pierce my mask is 

type(masks)->numpy.ndarray
im getting this output if i print it

```

masks
array([[[False, False, False, ..., False, False, False],
        [False, False, False, ..., False, False, False],
        [False, False, False, ..., False, False, False],
        ...,
        [False, False, False, ..., False, False, False],
        [False, False, False, ..., False, False, False],
        [False, False, False, ..., False, False, False]]])
```
```
passing this to function giving me error,


# Find contours in the mask image
contours, _ = cv2.findContours(masks, cv2.RETR_EXTERNAL, cv2.CHAIN_APPROX_NONE)
_, contours, _ = cv2.findContours(masks, cv2.RETR_EXTERNAL, cv2.CHAIN_APPROX_NONE)


# Convert the contour to the format required for segmentation in COCO format
segmentation = []
for contour in contours:
    # Flatten the contour coordinates into a 1D array
    contour = contour.flatten().tolist()
    # Convert the 1D array to a list of (x, y) pairs
    contour_pairs = [(contour[i], contour[i+1]) for i in range(0, len(contour), 2)]
    # Convert the (x, y) pairs to a list of integers
    segmentation.append([int(coord) for pair in contour_pairs for coord in pair])

# Define annotation in COCO format
annotation = {
    'id': 1,
    'image_id': 1,
    'category_id': 1,
    'segmentation': segmentation,
    'area': int(cv2.contourArea(contours[0])),
    'bbox': [int(x) for x in cv2.boundingRect(contours[0])],
    'iscrowd': 0
}
```






```

ERROR:

      1 # Find contours in the mask image
      2 #contours, _ = cv2.findContours(masks, cv2.RETR_EXTERNAL, cv2.CHAIN_APPROX_NONE)
----> 3 _, contours, _ = cv2.findContours(masks, cv2.RETR_EXTERNAL, cv2.CHAIN_APPROX_NONE)
      4 
      5 

error: OpenCV(4.7.0) :-1: error: (-5:Bad argument) in function 'findContours'
> Overload resolution failed:
>  - image data type = 0 is not supported
>  - Expected Ptr<cv::UMat> for argument 'image

```

**Comment by akashAD98** [comment]
```
import cv2
import matplotlib.pyplot as plt
import numpy as np
from ultralytics import YOLO
from segment_anything import sam_model_registry, SamPredictor

from PIL import Image


def yolov8_detection(image):
    model = YOLO('/content/best.pt')
    results = model(image, stream=True)  # generator of Results objects
    for result in results:
        boxes = result.boxes  # Boxes object for bbox outputs
    bbox = boxes.xyxy.tolist()[0]
    return bbox


def show_mask(mask, ax, random_color=False):
    if random_color:
        color = np.concatenate([np.random.random(3), np.array([0.6])], axis=0)
    else:
        color = np.array([30/255, 144/255, 255/255, 0.6])
    h, w = mask.shape[-2:]
    mask_image = mask.reshape(h, w, 1) * color.reshape(1, 1, -1)
    ax.imshow(mask_image)


def show_points(coords, labels, ax, marker_size=375):
    pos_points = coords[labels==1]
    neg_points = coords[labels==0]
    ax.scatter(pos_points[:, 0], pos_points[:, 1], color='green', marker='*', s=marker_size, edgecolor='white', linewidth=1.25)
    ax.scatter(neg_points[:, 0], neg_points[:, 1], color='red', marker='*', s=marker_size, edgecolor='white', linewidth=1.25)


def show_box(box, ax):
    x0, y0 = box[0], box[1]
    w, h = box[2] - box[0], box[3] - box[1]
    ax.add_patch(plt.Rectangle((x0, y0), w, h, edgecolor='green', facecolor=(0,0,0,0), lw=2))


# Load image
# image_path = '/content/myimg.jpg'
# image = cv2.imread(image_path)
# image = cv2.cvtColor(image, cv2.COLOR_BGR2RGB)


# Load image
image_path = 'Cigaretee_i_Weapon_Pragati137.jpg'
image = Image.open(image_path).convert('RGB')


# Detection with YOLOv5
yolov8_boxex = yolov8_detection(image)
input_box = np.array(yolov8_boxex)

# Segment with SAM
sam_checkpoint = "sam_vit_h_4b8939.pth"
model_type = "vit_h"
device = "cpu"
sam = sam_model_registry[model_type](checkpoint=sam_checkpoint)
sam.to(device=device)
predictor = SamPredictor(sam)
predictor.set_image(image)

masks, _, _ = predictor.predict(
    point_coords=None,
    point_labels=None,
    box=input_box[None, :],
    multimask_output=False,
)

```

**Comment by Jordan-Pierce** [comment]
@akashAD98 sorry, I should have said it's a non-tested response, but that's an approach that should help you get started. Here's what I'm using, hope it helps. 

```python

    # Read the image
    image = cv2.imread(image_path)
    image = cv2.cvtColor(image, cv2.COLOR_BGR2RGB)

    # Convert the bounding box to a tensor
    input_box = torch.tensor(bbox, device=predictor.device)
    transformed_box = predictor.transform.apply_boxes_torch(input_box, image.shape[:2])

    # Set the image
    predictor.set_image(image)

    # Predict the mask
    mask, _, _ = predictor.predict_torch(
        point_coords=None,
        point_labels=None,
        boxes=transformed_box,
        multimask_output=False,
    )   

    # Convert the mask to a binary image
    binary_mask = mask.cpu().numpy().squeeze().astype(np.uint8)

    # Find the contours of the mask
    contours, hierarchy = cv2.findContours(binary_mask,
                                        cv2.RETR_EXTERNAL,
                                        cv2.CHAIN_APPROX_SIMPLE)

    # Get the largest contour based on area
    largest_contour = max(contours, key=cv2.contourArea)

    # Get the new bounding box
    bbox = [int(x) for x in cv2.boundingRect(largest_contour)]

    # Get the segmentation mask for object 
    segmentation = largest_contour.flatten().tolist()

    return bbox, segmentation
```

**Comment by akashAD98** [comment]
@Jordan-Pierce than you so much. here is the implementation of yolov8 with SAM. if someone wants to try
https://github.com/akashAD98/YOLOV8_SAM

**Comment by akashAD98** [comment]
> We added support in Roboflow for [annotating using SAM](https://blog.roboflow.com/label-data-segment-anything-model-sam/). We convert the masks to polygons & you can export to YOLO format for training a model.

hi im opening my files in roboflow & its showing these extra lines, may i know what's wrong with this


<img width="151" alt="image" src="https://user-images.githubusercontent.com/62583018/232742456-37f527db-f348-4d01-ade2-fe513bba5536.png">


<img width="181" alt="image" src="https://user-images.githubusercontent.com/62583018/232742593-b0f8f77a-353f-420b-8d55-72003d984a42.png">

**Comment by yeldarby** [comment]
Looks like you’re not converting to polygons very well. You have points that lie outside of your core shape, your shape doesn’t start and end near the same spot, and you’re adding a polygon for every point along the perimeter of the mask. Looks like your coordinate space might be messed up too.

**Comment by yeldarby** [comment]
I don't know how to do it in Python; as noted above, [we built SAM into Roboflow](https://blog.roboflow.com/label-data-segment-anything-model-sam/) & [you can export your dataset in YOLO format](https://blog.roboflow.com/how-to-train-yolov8-on-a-custom-dataset/).

**Comment by JordaoCassiano** [comment]
How can I convert the masks file in .json format (COCO RLE) to a polygon format that can be read by an annotation software such as CVAT?

**Comment by akashAD98** [comment]
im done with script, if someone want to try they can feel free to use it.
https://github.com/akashAD98/YOLOV8_SAM

**Comment by SzaremehrjardiMT** [comment]
is there any solution to convert SA model output to Yolo segmentation format?

**Comment by akashAD98** [comment]
please check here https://github.com/akashAD98/YOLOV8_SAM @SzaremehrjardiMT

**Comment by ryouchinsa** [comment]
To convert the mask image to polygons, you can use this script [mask2polygons.py](https://github.com/ryouchinsa/Rectlabel-support/blob/master/mask2polygons.py).

To convert the COCO RLE format to polygons, you can use this script [rle2polygons.py](https://github.com/ryouchinsa/Rectlabel-support/blob/master/rle2polygons.py).

[RectLabel](https://rectlabel.com) is an offline image annotation tool for object detection and segmentation.
Although this is not an open source program, you can label polygons using Segment Anything models and read/write/export the YOLO segmentation format.

![sam_polygon](https://github.com/ryouchinsa/sam-cpp-macos/assets/1954306/4640e139-c533-4b8c-b27b-e02a401b9bbd)

**Comment by ryouchinsa** [comment]
Improved "Create polygon using SAM" feature so that you can label pixels using the pixels option. You can label [Segment Anything 1 Billion (SA-1B)](https://ai.meta.com/datasets/segment-anything-downloads/) like dataset by yourself.

![sam-pixels](https://github.com/facebookresearch/segment-anything/assets/1954306/d828c18b-6fda-4258-b866-21bc34a0d17f)

**Comment by sriram-dsl** [comment]
the easiest way to automate the labeling is 
follow this colab notebook : [colab](https://www.youtube.com/redirect?event=video_description&redir_token=QUFFLUhqbUpCc1pqRGNhWlpjU2ZZX3ZqN0ZYY3QyVGVid3xBQ3Jtc0trQmk0MW13YkI3LVUyTEs1MFdXdFpZZVJhT1l1Yzl4dW1OVE5KZUxHaHR5S1hMVkJ3dGU5TkpYZ0h2US1lbktJenp2OEVTZV9sb0t0el95R3M1eFUtZkg3MFpyeXJ3Umk2LUkwYzNPdE1MMDlZb2xJZw&q=https%3A%2F%2Fcolab.research.google.com%2Fgithub%2Froboflow-ai%2Fnotebooks%2Fblob%2Fmain%2Fnotebooks%2Fautomated-dataset-annotation-and-evaluation-with-grounding-dino-and-sam.ipynb&v=oEQYStnF2l8) 

then using dino you will get the labels in xml format
then convert the xml labels to yolo format using any open source frameworks
you can use makesense.ai to visualise and convert the labels to other formats
